[ErrorHandler] Render template with status code assigned to the exception

aderiyenko opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4 
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #44433 
| License       | MIT
| Doc PR        | N/A


* Injecting `framework.exceptions` config into `Symfony\Component\ErrorHandler\ErrorRenderer\HtmlErrorRenderer`
* Passing the status code from the config into `FlattenException::createFromThrowable()`
* Tests

GromNaN commented on src/Symfony/Component/ErrorHandler/ErrorRenderer/HtmlErrorRenderer.php:
```diff
@@ -63,22 +65,29 @@ public function __construct($debug = false, string $charset = null, $fileLinkFor
+        $statusCode = 500;
+        foreach ($this->exceptionsMapping as $exception => $config) {
+            if ($throwable instanceof $exception && $config['status_code']) {
+                $statusCode = $config['status_code'];
```
A `break` of the loop is necessary to stop at the first match.
Like https://github.com/symfony/symfony/blob/v5.4.0/src/Symfony/Component/HttpKernel/EventListener/ErrorListener.php#L91-L96

aderiyenko replied:
done

GromNaN commented on src/Symfony/Component/ErrorHandler/Tests/ErrorRenderer/HtmlErrorRendererTest.php:
```diff
@@ -53,5 +69,27 @@ public function getRenderData(): iterable
+        yield '->render() returns the HTML content WITH stack traces in debug mode and contains the correct status code' => [
+            new \RuntimeException('Foo'),
+            new HtmlErrorRenderer(true, null, null, null, '', null, [
+                  '\\RuntimeException' => [
```
We prefer the syntax `\RuntimeException::class` to help the IDE.

aderiyenko replied:
done

GromNaN commented on src/Symfony/Component/ErrorHandler/Tests/ErrorRenderer/HtmlErrorRendererTest.php:
```diff
@@ -53,5 +69,27 @@ public function getRenderData(): iterable
+        yield '->render() returns the HTML content WITHOUT stack traces in non-debug mode and contains the correct status code' => [
+            new \RuntimeException('Foo'),
+            new HtmlErrorRenderer(false, null, null, null, '', null, [
+                  '\\RuntimeException' => [
```
Same here `\RuntimeException::class`.

aderiyenko replied:
done, thanks

GromNaN reviewed (changes requested):
Thank you for fixing this bug. Some small changes to make it ready.

aderiyenko commented:
@GromNaN Thanks for the review! I'm not sure how I can fix the tests in the pipeline. Where can I read more about `Error: KO src/Symfony/Bundle/FrameworkBundle` ?

Also, it throws some linting error about extra return in phpdoc, but this is not related to my change, so I put it back.

GromNaN commented:
> @GromNaN Thanks for the review! I'm not sure how I can fix the tests in the pipeline. Where can I read more about `Error: KO src/Symfony/Bundle/FrameworkBundle` ?

This error a not related to your changes. There is the same errors on the target branch (5.4).

> Also, it throws some linting error about extra return in phpdoc, but this is not related to my change, so I put it back.

You can apply the patch from fabbot please.

aderiyenko commented:
done. I thought the pipeline should be all green in order to merge :) Anything else I can/should do in this branch?

GromNaN commented on src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRendererInterface.php:
```diff
@@ -23,5 +23,5 @@ interface ErrorRendererInterface
      * Renders a Throwable as a FlattenException.
      */
-    public function render(\Throwable $exception): FlattenException;
+    public function render(\Throwable $throwable): FlattenException;
```
I had to check to be sure, and the psalm errors confirmed: we don't rename arguments to avoid conflict when merging older branches into the newer. Even if I agree that this name would have been more accurate for a new code.

aderiyenko replied:
got it, will rename back in a few. Thanks!

aderiyenko replied:
done

GromNaN commented:
It would be perfect if you could squash your commits. (Personally I abuse the `--amend` option to fix commits)

aderiyenko commented:
squashed and set the commit message to contain the issue id

GromNaN commented on src/Symfony/Component/ErrorHandler/ErrorRenderer/HtmlErrorRenderer.php:
```diff
@@ -76,7 +79,14 @@ public function render(\Throwable $exception): FlattenException
 
-        $exception = FlattenException::createFromThrowable($exception, null, $headers);
+        $statusCode = 500;
+        foreach ($this->exceptionsMapping as $exc => $config) {
```
Last comment: we could use a more descriptive var name like `$class`.

aderiyenko replied:
done, thanks!

jderusse commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.php:
```diff
@@ -30,6 +30,7 @@
                     ->factory([HtmlErrorRenderer::class, 'getAndCleanOutputBuffer'])
                     ->args([service('request_stack')]),
                 service('logger')->nullOnInvalid(),
```
I don't understand the sentence. `an` + exception`s`. Did you mean`a list of exceptions`?
same for constructor docblock

aderiyenko replied:
I took it from [here](https://github.com/symfony/symfony/blob/6.1/src/Symfony/Bundle/FrameworkBundle/Resources/config/web.php#L108) and I also did not understand it first time when I read it, but now I think I understand it like "an (exceptions to log & status code) mapping", so the article refers to "mapping". Sorry, English is not my mother tongue, so I took it from the same parameter passed to the exception listener, but happy to change the wording. Question: if we decide to change, should I do that in the exception listener DI config as well?

GromNaN replied:
What do yo think of "Configuration per exception class"?

aderiyenko replied:
I'm fine with that description, but let me know if I should change the original usages of it as well

aderiyenko replied:
I mean for me it's hard to say now as I understand what it's all about and any of these descriptions is clear. However, for someone who sees it for the first time probably the latter one is better. But again, for me it's hard to say now (and I'm not a native English speaker)

GromNaN replied:
Same feeling for me.
You can update on both places. Thanks.

aderiyenko replied:
done, thanks!

yceruto commented:
For API exceptions it's the same problem.

@aderiyenko can you also update `SerializerErrorRenderer`? Thanks!

aderiyenko commented:
@yceruto sure, will look at it and try to do the same there and I saw your idea in the issue comments, so let me try to create another PR with another approach. I'll try to finish this one over the weekend and then will start with wrapping it into an http exception idea

aderiyenko commented:
@yceruto I added the same functionality to `SerializerErrorRenderer` and realized there were no tests for the class, so added a couple, but they test only the status code/text, i.e. this PR parts. Not testing whether stack trace is returned/hidden in debug/prod mode

aderiyenko commented:
changed the wording of the additional config (exception type -> log_level/status code) parameter everywhere and also changed the hardcoded constants (status codes and texts) in tests to use the ones from HttpFoundation\Response

Commit:
squashing

aderiyenko commented:
@GromNaN I see that my new tests pass in 7.4 pipeline, but fail in 8. See here: https://github.com/symfony/symfony/runs/4422885653?check_suite_focus=true . Problem is: I just tried to launch it inside a php:8.0.5-fpm-alpine container (with only xdebug3 installed) and it passes fine, so I can't recreate it locally. Is there any specific config in the pipeline that I can reproduce locally ?

aderiyenko commented:
Is my understanding correct that low- and high-deps pipelines install all components of certain versions, so the tests covering the new functionality in any PR will never work in those pipeline steps? Because these versions do not have the functionality being tested?

yceruto commented:
@aderiyenko you should update the affected composer.json accordingly to make it pass. For example, after these changes FrameworkBundle is no longer compatible with `"symfony/error-handler": "^5.4|^6.0"`, so you have to limit it to `^6.1`.

GromNaN commented:
@yceruto this is a bugfix, I think we should not update the dependencies in that case.

yceruto commented:
You're right, I confused the branch this PR targeted. Initially, it was 6.1.

wouterj commented:
Iirc, in that case, require the upcoming minor version: `^5.4.2|^6.0.2`

Commit:
bumping error handler version

aderiyenko commented:
Thanks everyone! I just bumped the versions of error handler in FrameworkBundle's composer.json , but the same errors are still there. But I'm more interested if there are certain steps I can take locally to reproduce the problem?

nicolas-grekas reviewed (changes requested):
Thank you for the PR, I really appreciate that you took the time to write and submit it.
Having reviewed the code, the added logic doesn't make sense to me. It feels like spreading a concern to places where it doesn't belong.
On a closer look, I think the implementation provided in #42244 is at fault and that it should be changed.
I gave it a try and submitted it at #44649.

Thanks again. I'm :-1: here as explained.

aderiyenko commented:
Sure, thanks @nicolas-grekas
